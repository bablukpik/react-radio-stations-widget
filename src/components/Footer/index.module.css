.footer {
  height: 160px;
  width: 100%;
  background-color: #22222b;
  border-top: 2px solid #42424c;
  border-bottom-left-radius: 47px;
  border-bottom-right-radius: 47px;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
}

.footerTitle {
  font-size: 22px;
  color: #edae61;
  font-weight: 300;
}
