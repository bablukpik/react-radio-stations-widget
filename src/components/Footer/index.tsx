import * as React from 'react'
import {State} from '../../context/stationsContext'
import {getCurrentlyPlayingStation} from '../../utils/helperFunctions'
import Text from '../Text'
import styles from './index.module.css'
interface Props {
  state: State
}

const Footer: React.FC<Props> = ({ state }) => {
  console.log('state', state)
  const [station] = getCurrentlyPlayingStation(state.currentlyPlaying, state)
  console.log('currently playing station', station);

  return (
    <footer className={styles.footer}>
      {station && (
        <>
          <Text className={styles.footerTitle}>CURRENTLY PLAYING</Text>
          <Text>{station.name}</Text>
        </>
      )}
    </footer>
  )
}

export default Footer
