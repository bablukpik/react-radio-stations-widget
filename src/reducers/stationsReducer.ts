export interface StationItem {
  name: string;
  id: string;
  isOpen: boolean;
}

export interface State {
  stations: Array<StationItem>
  currentlyPlaying: StationItem | null;
}

interface ActionTypes {
  setCurrentlyPlaying: string
  setStations: string
}

export type Action = {
  type: string;
  payload: {
    stations?: StationItem[];
    station?: StationItem;
  };
}

const initialState: State = {
  stations: [
    {name: 'Putin FM', id: '66,6', isOpen: false},
    {name: 'Dribbble FM', id: '101,2', isOpen: false},
    {name: 'Doge FM', id: '99,4', isOpen: false},
    {name: 'Ballads FM', id: '87,1', isOpen: false},
    {name: 'Maximum FM', id: '142,2', isOpen: false},
  ],
  currentlyPlaying: null,
}

const actionTypes: ActionTypes = {
  setStations: 'SET_STATIONS',
  setCurrentlyPlaying: 'SET_CURRENTLY_PLAYING',
}

export default function stationsReducer(state: State = initialState, action: Action) {
  switch (action.type) {
    case actionTypes.setStations:    
      return {
        ...state,
        stations: action?.payload?.stations,
      };
    case actionTypes.setCurrentlyPlaying:
      const copyStations = state.stations;

      const currentlyPlayingStation = copyStations.find(item => item.isOpen) as StationItem;
      if (currentlyPlayingStation 
        && currentlyPlayingStation?.id !== action.payload.station?.id
      ) {
        currentlyPlayingStation.isOpen = false;
        copyStations[copyStations.findIndex(el => el.id === currentlyPlayingStation?.id)] = currentlyPlayingStation;
      }

      const objIndex = copyStations.findIndex(((obj: StationItem) => obj.id === action.payload?.station?.id));
      copyStations[objIndex].isOpen = action?.payload?.station?.isOpen as boolean;
      return {
        stations: copyStations,
      };
    default:
      return state
  }
}
