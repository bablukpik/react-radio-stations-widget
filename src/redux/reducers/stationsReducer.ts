import { setCurrentlyPlaying } from '../../utils/helperFunctions';

export interface StationItems {
  name: string
  number: string
  isOpen: boolean
}

export interface State {
  stations: Array<StationItems>
  currentlyPlaying: string | null
}

interface ActionTypes {
  setCurrentlyPlaying: string
  getStations: string
  updateStations: string
}

export type Action = {
  type: string
  payload: string | boolean | Array<StationItems> | null
}

const initialState: State = {
  stations: [
    {name: 'Putin FM', number: '66,6', isOpen: false},
    {name: 'Dribbble FM', number: '101,2', isOpen: false},
    {name: 'Doge FM', number: '99,4', isOpen: false},
    {name: 'Ballads FM', number: '87,1', isOpen: false},
    {name: 'Maximum FM', number: '142,2', isOpen: false},
  ],
  currentlyPlaying: null,
}

const actionTypes: ActionTypes = {
  getStations: 'GET_STATIONS',
  updateStations: 'UPDATE_STATIONS',
  setCurrentlyPlaying: 'SET_CURRENTLY_PLAYING',
}

export default function stationsReducer(state = initialState, action: Action) {
  switch (action.type) {
    case actionTypes.getStations:
      return state;
    case actionTypes.setCurrentlyPlaying:
      if (state.currentlyPlaying === action.payload) {
        return state;
      }
      return {
        ...state,
        stations: setCurrentlyPlaying(state, action),
        currentlyPlaying: action.payload,
      };
    default:
      return state
  }
}
