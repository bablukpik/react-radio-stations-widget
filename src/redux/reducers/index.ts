import { combineReducers } from "redux";
import stationsReducer from './stationsReducer';
const reducers = combineReducers({
  allStations: stationsReducer,
});
export default reducers;
