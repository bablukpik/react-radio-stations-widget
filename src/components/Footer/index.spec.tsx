import { render } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { initialState, State } from 'src/reducers/stationsReducer';

import Footer from './index';

const getProps = (newState: State = initialState) => ({ 
    state: newState,
  });

describe('Footer', () => {
  it('should render', () => {
    const { asFragment } = render(<Footer {...getProps()} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders footer text', () => {
    const newState = {
        stations: [
          {name: 'Putin FM', id: '66,6', isOpen: true},
        ],
      };

    const { getByText } = render(<Footer {...getProps(newState)}/>);
    expect(getByText('Currently playing')).toBeInTheDocument();
    expect(getByText('Putin FM')).toBeInTheDocument();
  });
});
