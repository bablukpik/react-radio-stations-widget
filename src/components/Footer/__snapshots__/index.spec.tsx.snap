// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Footer should render 1`] = `
<DocumentFragment>
  <footer
    class="footer"
  />
</DocumentFragment>
`;
